### PR TITLE
Expose config file locations as env vars

### LINKED ISSUE #356: Expose config file locations as env vars
In the [factorio-server-charts](https://github.com/SQLJames/factorio-server-charts), to allow us to specify the config values we are using configmaps, which creates the json file and mounts it to the proper location, however they are Read only files. This poses an issue when the server-id.json is needed to create a public server because it needs rw operations.  I am currently looking into creating a sidecar to copy the files into the proper directory so we can properly rw to the directory. 

But in my digging to resolve this issue, this solution did come up as a potential fix, so I had to ask if your project would be open to accepting a pull request that allows users to specify the location of config files as an ENV variable.

For example, in the docker-entrypoint.sh file, we could check to see if the ENV is specified and if so, we would use that value instead of the location put in the flag arg list. For example:

```
#-z string True if the string is null
if [ -z "${FACTORIO_SERVER_ID_LOCATION}" ]; then
    FACTORIO_SERVER_ID_LOCATION="/factorio/config/server-id.json"
fi

FLAGS=(\
  --port "$PORT" \
  --server-settings "$CONFIG/server-settings.json" \
  --server-banlist "$CONFIG/server-banlist.json" \
  --rcon-port "$RCON_PORT" \
  --server-whitelist "$CONFIG/server-whitelist.json" \
  --use-server-whitelist \
  --server-adminlist "$CONFIG/server-adminlist.json" \
  --rcon-password "$(cat "$CONFIG/rconpw")" \
  --server-id $FACTORIO_SERVER_ID_LOCATION \
)
```

### PR DESCRIPTION
**Pull request by SQLJames**
Fixes #356 

In the [factorio-server-charts](https://github.com/SQLJames/factorio-server-charts), to allow us to specify the config values we are using configmaps, which creates the json file and mounts it to the proper location, however they are Read only files. This poses an issue when the server-id.json is needed to create a public server because it needs rw operations.  I am currently looking into creating a sidecar to copy the files into the proper directory so we can properly rw to the directory. 

But in my digging to resolve this issue, this solution did come up as a potential fix, so I had to ask if your project would be open to accepting a pull request that allows users to specify the location of config files as an ENV variable.

### CODE REVIEW

**Comment by SQLJames** [comment]
Fixes #356

**Comment by SQLJames** [comment]
I am actually not sure why the automated tests are failing, can someone tell me what is wrong? I was able to build these locally without issue.

**Comment by SuperSandro2000** [comment]
Because the script is broken and missing a -t.

File: 0.16/files/docker-entrypoint.sh [hunk 1]
```diff
@@ -50,13 +50,34 @@ if ! find -L "$SAVES" -iname \*.zip -mindepth 1 -print | grep -q .; then
     --map-settings "$CONFIG/map-settings.json"
 fi
 
+#-z string True if the string is null
```
Comment (SuperSandro2000) [inline 0.16/files/docker-entrypoint.sh:53]: I would prefer if you did not update 0.16 or 0.17. I didn't touch the code in a longer while and don't want to break it. Also the comments are not required and this script is still sh not bash.

File: 0.18/files/docker-entrypoint.sh [hunk 2]
```diff
@@ -76,16 +76,42 @@ if [[ $GENERATE_NEW_SAVE == true ]]; then
     fi
 fi
 
+#-z string True if the string is null
```
Comment (SuperSandro2000) [inline 0.18/files/docker-entrypoint.sh:79]:
```suggestion
```

Not required. Please remove them.

File: 0.18/files/docker-entrypoint.sh [hunk 3]
```diff
@@ -76,16 +76,42 @@ if [[ $GENERATE_NEW_SAVE == true ]]; then
...
+    FACTORIO_SERVER_WHITELIST="$CONFIG/server-whitelist.json"
+    echo "Setting Default Factorio Server whitelist Location"
+fi
+
+#-z string True if the string is null
+if [[ -z "${FACTORIO_SERVER_BANLIST:-}" ]]; then
+    FACTORIO_SERVER_BANLIST="$CONFIG/server-banlist.json"
+    echo "Setting Default Factorio Server banlist Location"
+fi
+mkdir -p $(dirname "$FACTORIO_SERVER_ID_LOCATION")
```
Comment (SuperSandro2000) [inline 0.18/files/docker-entrypoint.sh:100]:
```suggestion
mkdir -p "$(dirname "$FACTORIO_SERVER_ID_LOCATION")"
```

File: 0.18/files/docker-entrypoint.sh [hunk 4]
```diff
@@ -76,16 +76,42 @@ if [[ $GENERATE_NEW_SAVE == true ]]; then
...
+mkdir -p $(dirname "$FACTORIO_SERVER_ADMIN_LIST")
```
Comment (SuperSandro2000) [inline 0.18/files/docker-entrypoint.sh:101]:
```suggestion
mkdir -p "$(dirname "$FACTORIO_SERVER_ADMIN_LIST")"
```

File: 0.18/files/docker-entrypoint.sh [hunk 5]
```diff
@@ -76,16 +76,42 @@ if [[ $GENERATE_NEW_SAVE == true ]]; then
...
+mkdir -p $(dirname "$FACTORIO_SERVER_WHITELIST")
```
Comment (SuperSandro2000) [inline 0.18/files/docker-entrypoint.sh:102]:
```suggestion
mkdir -p "$(dirname "$FACTORIO_SERVER_WHITELIST")"
```

File: 0.18/files/docker-entrypoint.sh [hunk 6]
```diff
@@ -76,16 +76,42 @@ if [[ $GENERATE_NEW_SAVE == true ]]; then
...
+mkdir -p $(dirname "$FACTORIO_SERVER_BANLIST")
```
Comment (SuperSandro2000) [inline 0.18/files/docker-entrypoint.sh:103]:
```suggestion
mkdir -p "$(dirname "$FACTORIO_SERVER_BANLIST")"
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
CI should be fixed on master and because I cannot push to your branch you need to rebase on master.

File: 0.16/files/docker-entrypoint.sh [hunk 7]
```diff
@@ -59,4 +59,4 @@ $SU_EXEC /opt/factorio/bin/x64/factorio \
   --rcon-port "$RCON_PORT" \
   --rcon-password "$(cat "$CONFIG/rconpw")" \
   --server-id /factorio/config/server-id.json \
-  "$@"
+  "$@"
```
Comment (SuperSandro2000) [inline 0.16/files/docker-entrypoint.sh:62]: Missing final new line.

File: 0.17/files/docker-entrypoint.sh [hunk 8]
```diff
@@ -95,4 +95,4 @@ else
 fi
 
 # shellcheck disable=SC2086
-exec $SU_EXEC /opt/factorio/bin/x64/factorio "${FLAGS[@]}" "$@"
+exec $SU_EXEC /opt/factorio/bin/x64/factorio "${FLAGS[@]}" "$@"
```
Comment (SuperSandro2000) [inline 0.17/files/docker-entrypoint.sh:98]: Missing final new line.

**Comment by SuperSandro2000** [comment]
Also please never merge master in a PR and squash the commits together.

**Comment by SQLJames** [comment]
> Also please never merge master in a PR and squash the commits together.

Sorry this was the problem with rebasing off master, I ran into this issue then my existing pull request was already pointing to master. But yes, I will note that for the future.

File: 0.16/files/docker-entrypoint.sh [hunk 9]
```diff
@@ -60,3 +60,4 @@ $SU_EXEC /opt/factorio/bin/x64/factorio \
   --rcon-password "$(cat "$CONFIG/rconpw")" \
   --server-id /factorio/config/server-id.json \
   "$@"
+
```
Comment (SuperSandro2000) [inline 0.16/files/docker-entrypoint.sh:63]:
```suggestion
```

**Comment by proegssilb** [comment]
What's the status of this? It looks like there's a failing check for current, but I can't see any logs, and I'm not sure why these changes would fail for one version but not another?

File: 1.1/files/docker-entrypoint.sh [hunk 10]
```diff
@@ -76,16 +76,41 @@ if [[ $GENERATE_NEW_SAVE == true ]]; then
     fi
 fi
 
+if [[ -z "${FACTORIO_SERVER_ID_LOCATION:-}" ]]; then
```
Comment (perobertson) [inline 1.1/files/docker-entrypoint.sh:79]: To set default values you could change this if block to

```bash
: ${FACTORIO_SERVER_ID_LOCATION:=/factorio/config/server-id.json}
```

Example:
```
: ${test:=1 2} && echo $test
1 2

: ${test:=3 4} && echo $test
1 2
```
The `:` is to make the command a nop, but the default value still gets set

**Comment by Fank** [comment]
@SQLJames we did a lot of changes and got rid of the version numbers in the folder structure.
Could you try to do you changes based on the current master?

**Comment by SQLJames** [comment]
solved on the helm chart. https://github.com/SQLJames/factorio-server-charts/pull/13

**Pull request closed without merging**